File: python/medium/Merge_Intervals.py
class Solution:
    def check_merge(self, answer: List[List[int]]) -> List[List[int]]:
        index = 0
        while len(answer) != index and index != len(answer)-1:
            if answer[index][1] >= answer[index+1][1]:
                if answer[index][0] <= answer[index+1][0]:
                    answer.pop(index+1)
                elif answer[index][0] > answer[index+1][1]:
                    t = answer[index]
                    answer[index] = answer[index+1]
                    answer[index+1] = t
                    index+=1
                else:
                    answer[index] = [answer[index+1][0], answer[index][1]]
            else:
                if answer[index][1] < answer[index+1][0]:
                    break
                elif answer[index][0] >= answer[index+1][0]:
                    answer.pop(index)
                else:
                    answer[index] = [answer[index][0], answer[index+1][1]]
        return answer

    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        answer = [intervals[0]]
        for interval in intervals[1:]:
            answer = [interval] + answer
            answer = self.check_merge(answer)
            print(answer)
        return answer
      
      
      
# class Solution:
#     def merge(self, intervals: List[List[int]]) -> List[List[int]]:
#         # O(nlogn)
        
#         # sort intervals, sort by first interval index
#         intervals.sort(key=lambda i:i[0])
        
#         # initialize with first interval
#         ans = [intervals[0]]

        
#         for first, second in intervals[1:]:
#             last_second = ans[-1][1]
       
#             if first <= last_second:
#                 ans[-1][1] = max(last_second, second)
#             else:
#                 ans.append([first, second])
        
#         return ans

File: cpp/medium/Merge_Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> answer;
        vector<int> interval;
        sort(intervals.begin(), intervals.end());
        answer.push_back(intervals[0]);
        for (int i = 1; i < intervals.size(); ++i){
            interval = intervals[i];
            while(answer.size() != 0 && interval[0] <= answer.back()[1]) {
                cout << answer.size() << endl;
                interval[0] = min(answer.back()[0], interval[0]);
                interval[1] = max(answer.back()[1], interval[1]);
                answer.pop_back();
            }
            answer.push_back(interval);
        }
        return answer;
    }
};
